### PR TITLE
Improve linear time model

### PR DESCRIPTION
**Pull request by Nathaniel-Haines**
@jeremy-haynes I am opening a PR that is to be merged into your own branch after you have a look over it. A few things that are updated: 

- the `igt_orl_joint_indMu.stan` model now uses an intercept-slope form such that the person-level parameters have an intercept that they start at on session 1, and then the slope indicates the change in parameter with each day between session 1 and 2. Basically, the pseudo-model for each parameter is: `Arew_person_session ~ Arew_int + day * Arew_slope`, where each subject gets their own`Arew_int` and `Arew_slope`. This model should still work when we add more sessions, but right now we just have two (so the slope simply indicates the change from session 1 to session 2).
- the `Preprocess1.R` code now has a section at the bottom that puts together the stan data for the joint model. You will see that I am computing the number of days between session 1 and 2 for each person, and then using that as the `day` predictor in the stan model. This file needs to be run first (you can just `source()` the whole file) before trying to fit the stan model. 

That's about it! Let me know what you think about the change—I am fitting the model now, will get back here with some preliminary results once they come in 🤓

### CODE REVIEW

**Comment by Nathaniel-Haines** [comment]
@jeremy-haynes As results are coming in, I am finding that convergence is rather poor—specifically, a few chains get stuck in a weird part of the parameter space, and they have many divergences, poorer fit, and take longer to sample than other well-behaved chains. I am going to tighten up the group-level priors and see if that improves convergence! It should given the pattern I am seeing, where the bad chains are in an extreme part of the parameters space. We shall see 🤓

**Comment by Nathaniel-Haines** [comment]
> I am going to tighten up the group-level priors and see if that improves convergence! It should given the pattern I am seeing, where the bad chains are in an extreme part of the parameters space. We shall see 🤓

Success! 

Screen Shot 2023-05-14 at 9 04 10 PM

And pairs plots: 

Screen Shot 2023-05-14 at 9 03 58 PM

The solution was to make the priors on the slope parameters tighter, which in hindsight is obvious. e.g., time predictor going into the model is not scaled, so it is around 28 days for most people (month between test and re-test). So, the change per day (which the slope parameter captures) should be near 0 in a realistic scenario where the parameters change over time, but not be a substantial amount. So, I used `slope ~ normal(0, .025)` rather than the default `slope ~ normal(0, 1)` to reflect this, and convergence is great now 😄

File: 1_IGT_PP/Code/R/0_Preprocess/Preprocess1.R
Comment (jeremy-haynes) [inline 1_IGT_PP/Code/R/0_Preprocess/Preprocess1.R:1]: Thank you for adding this!

**Review by jeremy-haynes — APPROVED** [review]
@Nathaniel-Haines this looks great! It all makes sense to me.

**Comment by Nathaniel-Haines** [comment]
@jeremy-haynes @hollysully for context, here is what the plots above looked like when I had both the group-level intercept and slope specified as `normal(0,1)`:

Screen Shot 2023-05-15 at 3 51 11 PM

Screen Shot 2023-05-15 at 3 53 48 PM

You can see that one chain (in this case, it was more when I ran before) is in a very different part of the parameter space. The `pairs` plots highlight transitions that go over the maximum tree depth in yellow, and divergent transition in red. You can see bimodality in the `pairs` plots, where all the transitions in the extreme parts are hitting max tree depth, and there are some divergences there too. This issue is mostly with `K`, but also shows up with `betaP`. 

This is happening because the `normal(0,1)` priors on the slope parameters are generally putting far too much probability mass on part of the parameter space that are completely unreasonable. As I noted above, with "days" being the unit that the slope parameter indicates, even a slope as small as .5 would translate to a `28 * .5 = 14` change in the parameter from session 1 to 2. For the learning model parameters, obviously this would not be realistic. 

I came up with `normal(0, .025)` by noting that most people have 28 days between sessions. We are assuming then that a `28*.025 = .7` change is within 1 SD of what we expect a priori, and `28*.05 = 1.4` within 2 SD of what we expect. This seems more reasonable to me—we still allow for extreme changes, but even the most extreme changes we expect are still changes that are, in principle, possible. 

Let me know if that makes sense! Happy to chat more here.

**Comment by jeremy-haynes** [comment]
@Nathaniel-Haines that does make sense... essentially adjusting the prior to restrict the MCMC to avoid sampling implausible values of the parameter (e.g., a slope of .5 as in your example). Cool!

**Comment by jeremy-haynes** [comment]
@Nathaniel-Haines @hollysully I do have a question. If we wanted to examine test-retest with this model, would we do it similar to how we've done before where we correlate session 1 and session 2 parameters at each iteration to construct a posterior distribution of the correlations (then apply the `pop_correct_MCMC` function as necessary)? Or would we have to do this in a different way (e.g., like an ICC)?

**Comment by hollysully** [comment]
> @Nathaniel-Haines @hollysully I do have a question. If we wanted to examine test-retest with this model, would we do it similar to how we've done before where we correlate session 1 and session 2 parameters at each iteration to construct a posterior distribution of the correlations (then apply the `pop_correct_MCMC` function as necessary)? Or would we have to do this in a different way (e.g., like an ICC)?

@Nathaniel-Haines will likely have more nuanced thoughts, but I don't see why we couldn't do more-or-less exactly what we did for the previous test-retest paper for this adapted model. I assume you would want to use the earlier joint model rather than this new longitudinal model (currently set up for two time points)......but comparing test-retest with the earlier joint model and this new longitudinal model (as well as a regular, old two-step approach) might also be cool!?

**Comment by Nathaniel-Haines** [comment]
> > @Nathaniel-Haines @hollysully I do have a question. If we wanted to examine test-retest with this model, would we do it similar to how we've done before where we correlate session 1 and session 2 parameters at each iteration to construct a posterior distribution of the correlations (then apply the `pop_correct_MCMC` function as necessary)? Or would we have to do this in a different way (e.g., like an ICC)?
> 
> @Nathaniel-Haines will likely have more nuanced thoughts, but I don't see why we couldn't do more-or-less exactly what we did for the previous test-retest paper for this adapted model. I assume you would want to use the earlier joint model rather than this new longitudinal model (currently set up for two time points)......but comparing test-retest with the earlier joint model and this new longitudinal model (as well as a regular, old two-step approach) might also be cool!?

@jeremy-haynes I agree with @hollysully! We should be able to just use the same method as before, which makes things easier. We could also probably compute the correlation using some semi-analytical method that takes into account the variance+covariance of slopes and intercepts, but I think it is probably best to just use the same method as before. At least then we can be sure that any differences between the estimated correlation between models is due to the model specification and not how we obtain the correlation 😄 

Also @jeremy-haynes , we can merge this into your PR now! And then merge yours into `main` if you would like to. Happy to help walk through that process if it would be useful 🤓

**Comment by jeremy-haynes** [comment]
@Nathaniel-Haines @hollysully Cool thank you! That makes sense. I added a few things to do that in a recent commit I just pushed ([here](https://github.com/hollysully/IGT_PP_Shared/commit/33b86725edf69ef0cab9e6f2a37e7e46f3aeec15)) has the correlations. The correlations are surprisingly higher than the regular joint model for each parameter except for Arew (see table below). I am surprised of how the correlation between timepoints with K increased. Might be worth double-checking that I'm not missing something.
image

Also, should I go ahead and run this with our usual setup for the MCMC (4 chains with 5000 iterations and 1000 warmup)?

**Comment by jeremy-haynes** [comment]
Oh, and @Nathaniel-Haines that sounds good about merging the PRs. Walking through it would be helpful. Is that something I can do on the terminal or is that a GitHub online thing?

**Pull request merged**